examples/simple_fwd_tb/forward_tb.c

```diff
@@ -0,0 +1,324 @@
+                        tb_tokens = tb_depth;
+                } else {
+                        tb_tokens += tokens_produced;
+                }
+
+                last_cycle = cur_cycles;
+        }
+
+        if (++counter == print_delay) {
+                do_stats_display(pkt);
+                counter = 0;
+        }
+
+        tb_tokens -= pkt->pkt_len;
+
+        meta->action = ONVM_NF_ACTION_TONF;
+        meta->destination = destination;
+        return 0;
+}
+
+void
+sig_handler(int sig) {
+        if (sig != SIGINT && sig != SIGTERM)
+                return;
+
+        /* Will stop the processing for all spawned threads in advanced rings mode */
+        rte_atomic16_set(&signal_exit_flag, 1);
+}
+
+static int
+thread_main_loop(struct onvm_nf_local_ctx *nf_local_ctx) {
+        void *pkts[PKT_READ_SIZE];
+        struct onvm_pkt_meta *meta;
+        uint16_t i, nb_pkts;
+        void *pktsTX[PKT_READ_SIZE];
+        int tx_batch_size;
+        struct rte_ring *rx_ring;
+        struct rte_ring *tx_ring;
+        struct rte_ring *msg_q;
+        struct onvm_nf *nf;
+        struct onvm_nf_msg *msg;
+        struct rte_mempool *nf_msg_pool;
+
+        nf = nf_local_ctx->nf;
+
+        onvm_nflib_nf_ready(nf);
+
+        /* Get rings from nflib */
+        rx_ring = nf->rx_q;
+        tx_ring = nf->tx_q;
+        msg_q = nf->msg_q;
+        nf_msg_pool = rte_mempool_lookup(_NF_MSG_POOL_NAME);
+
+        printf("Process %d handling packets using advanced rings\n", nf->instance_id);
+        if (onvm_threading_core_affinitize(nf->thread_info.core) < 0)
+                rte_exit(EXIT_FAILURE, "Failed to affinitize to core %d\n", nf->thread_info.core);
+
+        last_cycle = rte_get_tsc_cycles();
+
+        while (!rte_atomic16_read(&signal_exit_flag)) {
+                /* Check for a stop message from the manager */
+                if (unlikely(rte_ring_count(msg_q) > 0)) {
+                        msg = NULL;
+                        rte_ring_dequeue(msg_q, (void **)(&msg));
+                        if (msg->msg_type == MSG_STOP) {
+                                rte_atomic16_set(&signal_exit_flag, 1);
+                        } else {
+                                printf("Received message %d, ignoring", msg->msg_type);
+                        }
+                        rte_mempool_put(nf_msg_pool, (void *)msg);
+                }
+
+                tx_batch_size = 0;
+                nb_pkts = rte_ring_dequeue_burst(rx_ring, pkts, PKT_READ_SIZE, NULL);
+
+                /* Process all the dequeued packets */
+                for (i = 0; i < nb_pkts; i++) {
+                        meta = onvm_get_pkt_meta((struct rte_mbuf *)pkts[i]);
+                        packet_handler_tb((struct rte_mbuf *)pkts[i], meta, nf_local_ctx);
+                        pktsTX[tx_batch_size++] = pkts[i];
+                }
+
+                rte_ring_enqueue_bulk(tx_ring, pktsTX, tx_batch_size, NULL);
```

Comment:
Please implement the new fix to advanced rings listed here: https://github.com/sdnfv/openNetVM/pull/212 this should fix any stats bugs

Reply:
Right now, the code works in bursts of 32 itself so I don't think there's any problem, but will update the code just to be safe as soon as the #212 get's approved/merged.